homeassistant/components/statistics/sensor.py

```diff
@@ -281,22 +281,22 @@ def _next_to_purge_timestamp(self):
 
         # If max_age is set, ensure to update again after the defined interval.
         next_to_purge_timestamp = self._next_to_purge_timestamp()
-        if next_to_purge_timestamp:
+        if (
+            next_to_purge_timestamp is not None
+            and (next_to_purge_timestamp - self._next_purge).total_seconds() > 1
```

Comment:
Are we now assuming that an already scheduled callback will have run within 1 second? What if that doesn't happen, maybe due to some congestion? Shouldn't we still make sure that there's only one scheduled callback at any one time?

Reply:
> What **if that doesn't happen**, maybe due to some congestion?

Well, the thing is that it seems the congestion was caused by the old way of scheduling callbacks for each sample, but I didn't think about that case.. is that possible?

The intention was to avoid too many callbacks for samples that come almost at the same time:

```
S1(t=0) --> P1(t=3)
S2(t=1) --> P2(t=4)
S3(t=2) --> P3(t=5)
S4(t=2.001) --> No scheduled purge
S5(t=2.01) --> No scheduled purge
S6(t=2.5) --> No scheduled purge
S7(t=3) --> P7(t=6)

# where Sx is a new sample and Px is the scheduled purge call, with t in seconds and max_age=3s
```

> Shouldn't we still make sure that there's only one scheduled callback at any one time?

I don't understand this, sorry. At any time `t` there are `n` scheduled callbacks in the future (for different t+x times), created by the last n samples in the past. If a new sample n+1 is received, we program a new callback for t+max_age, **but only if t - t_last > 1.

If there is congestion in the event bus, creating more callbacks is not the best solution, right?

What I didn't review was the **purge process**, to make sure that, in the example, `P3` does remove S3,S4,S5 and S6, but no S7. If not, the fix is a bad one.

I think I'm going to try to write some test to show this better, what do you think @MartinHjelmare ?

Reply:
Also, the variable naming is bad, as I see now: `self._next_purge` should be named **`self._last_scheduled_purge`** or similar

Reply:
I'm not too familiar with this integration, that's why I thought we only needed one callback at any one time.

Reply:
Do we really need more than one callback scheduled at any one time? The next purge time is taken from the oldest stored state. The oldest stored state is only changed when purging. So then we only need one callback. Am I missing something?

Reply:
It seems the deques of states and ages have a max length, so the oldest item might also change if we exceed this length when adding items.

Reply:
> I'm not too familiar with this integration

Also my case :)

> Do we really need more than one callback scheduled at any one time? The next purge time is **taken from the oldest stored state**.

Oops, totally missed that, you're right. The schema I pointed before is not good, but I think my point remains, more or less.

> Am I missing something?

Looks like I was the one :)

> It seems the deques of states and ages have a max length, so the oldest item might also change if we exceed this length when adding items.

Yes, but this fact only affects if the sensor has a `max_age` that saturates the `sampling_size` (sometimes). I think is in those situations when the issue appears, but I'm not sure.

When the oldest item changes and the new one only differs in milliseconds from the old one, a new scheduled callback is created, and then in the cancel process for the old one, _someway_, _sometimes_, the core warning appears.

The thing is that **the code before this PR should work**, as far as I can see it, **but it didn't** completely, and when `self._update_listener()` was called, sometimes it was not recognized as a known listener.

My point in this PR is to do not stress the listener pipe by creating and canceling so many listeners (even when, as you pointed right, there is only one scheduled at each t)

I have 2 more commits on another branch, which was waiting for another PR with a new feature for this: de-coupling the state update from the sampling process, and make it at a fixed scan_interval frequency (optionally).

Now async_update (which does the purging, all the sample processing, and the scheduling) is called for each received sample, which can be A LOT
Decoupling it can reduce the number of state changes for this kind of sensors (which is something interesting for me and I think for more people)

Reply:
I think the aim with less scheduling sounds good. But we need to make sure we don't make false assumptions.

Maybe we should ask for some more eyes on this PR?

@balloob?

Reply:
Yes, totally. I think @exxamalte is who has more context here.

Reply:
Alright: The original intention was to always only have a single listener for each configured integration. This listener is instantiated with the timestamp when the next purge is supposed to happen. When this time is reached, `async_update` is called, and the first thing that happens there is that all old values are purged before the statistics sensor state is updated. At the end of the update, the next timestamp to purge is determined and another listener is created.
Now, there are 2 ways that `async_update` can be called - either because an observed entity has changed its state  or because the time to purge is reached.

This PR now aims to reduces the number of listeners:
**Before/Now:**
1. Determine next time to purge.
2. Remove old listener.
3. Create new listener.
**Future:**
1. Determine next time to purge.
2a. If that next time is the same as or later (=further in the future) than in last `async_update` call: Do Nothing (i.e. keep old listener in place).
2b. If that next time is earlier than in the last `async_update` call: Remove old listener, and create new listener.

```
and (next_to_purge_timestamp - self._next_purge).total_seconds() > 1
```
Where `next_to_purge_timestamp` is the very next timestamp to purge and to schedule the next update, and `self._next_purge` is the timestamp from the last time an update was conducted. This part evaluates `true` only if `next_to_purge_timestamp` a bit is larger than `self._next_purge`, i.e. it's not correct to then create a new listener. I think this should rather be `(next_to_purge_timestamp - self._next_purge).total_seconds() <= 0`.

Reply:
Should we use `helpers/debounce.py` to debounce the call properly?

Reply:
Looking at this code, I think that the fix should be that `self._next_to_purge_timestamp()` should do the check instead of having to check the return value because it's not actually returning the proper next to purge timestamp.

Reply:
> Should we use `helpers/debounce.py` to debounce the call properly?

That sounds like a good option. However, I believe that this should then not happen in this component, but in the underlying `async_track_point_in_utc_time`. Otherwise, we may run into the same issue in other components that use this function.

My view might be a bit biased because I implemented the original change in the `statistics` sensor, but I still believe that I had originally used the correct way to create and remove listeners.

`async_track_point_in_utc_time` returns a listener, but apparently it can happen that when a component then calls that listener to remove it, an update of that same listener may already be in progress leading to the original issue that we have seen (https://github.com/home-assistant/home-assistant/issues/31576).

Reply:
> I think that the fix should be that `self._next_to_purge_timestamp()` should do the check instead of having to check the return value because it's not actually returning the proper next to purge timestamp.

At the moment, `_next_to_purge_timestamp()` is calculating the exact timestamp when the next purge is supposed to happen (or in other words: when the oldest observed entity's state is going to expire), and thus when the next update should be scheduled.

Reply:
I think I introduced more confusion than clearness here :(
I'll try to re-enter:

@exxamalte,

> Determine next time to purge.
2a. If that next time is the same as or later (=further in the future) than in last async_update call: Do Nothing (i.e. keep old listener in place).
2b. If that next time is earlier than in the last async_update call: Remove old listener, and create new listener.

As max_age does not change, the next time to purge has to be >= than the last one, so (2b) cannot happen. New times are **the same** or greater.

The problem resides in the moments that it is the same as last one (because 2 calls to async_update are made at ~ the same time, usually because 2 samples are added)

My initial intention was to **filter** the scheduling when the last one was too similar, to avoid this:

```log
2020-02-23 23:56:01 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update
2020-02-23 23:56:01 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=66] with next purge at: 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:01 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:01 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=67] with next purge at: 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:01 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:01 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: Cancelling old listener
2020-02-23 23:56:02 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=68] with next purge at: 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:02 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 22:56:02.010940+00:00
2020-02-23 23:56:02 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: Cancelling old listener
2020-02-23 23:56:03 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update
2020-02-23 23:56:03 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=67] with next purge at: 2020-02-23 22:56:04.007351+00:00
2020-02-23 23:56:03 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 22:56:04.007351+00:00
2020-02-23 23:56:04 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=68] with next purge at: 2020-02-23 22:56:04.007351+00:00
2020-02-23 23:56:04 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 22:56:04.007351+00:00
2020-02-23 23:56:04 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: Cancelling old listener
```

As you can see, incoming samples at the same time make repeated cancel-program_again cycles with the scheduled listener, and that is (I think) what is making the mess when this situation is much more chaotic.

By filtering this scheduling when `(next_to_purge_timestamp - self._next_purge).total_seconds() > 1` we avoid the cancel + program again cycle:

```log
2020-02-24 00:00:05 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update
2020-02-24 00:00:05 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=69] with next purge at: 2020-02-23 23:00:06.008686+00:00
2020-02-24 00:00:05 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 23:00:06.008686+00:00
2020-02-24 00:00:05 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=70] with next purge at: 2020-02-23 23:00:06.008686+00:00
2020-02-24 00:00:06 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=71] with next purge at: 2020-02-23 23:00:06.008686+00:00
2020-02-24 00:00:07 WARNING (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: executing scheduled update
2020-02-24 00:00:07 INFO (MainThread) [homeassistant.components.statistics.sensor] Updating sensor.stats_1 [n=70] with next purge at: 2020-02-23 23:00:08.007817+00:00
2020-02-24 00:00:07 INFO (MainThread) [homeassistant.components.statistics.sensor] sensor.stats_1: scheduling update at 2020-02-23 23:00:08.007817+00:00
```

But I removed the storage and canceling logic for the listener and that was bad, I'll revert it.

Anyway, as I commented in the conversation in the issue, IMO the root problem is the high number of calls to update state, and about that:

cc @balloob 

> Should we use helpers/debounce.py to debounce the call properly?

Didn't know about that nice helper :)
Definitely I think this is a bouncing problem of the state updater, how to use that to limit the nº of calls to `async_update`?

Should we just _debounce_ the call to update done after each received sample?

```
self._add_state_to_queue(new_state)
# debounce this call:
self.async_schedule_update_ha_state(True)
```

I will fix the things I broke here, but introduce the filtering anyway, as it really removes the issue: I have the RPI running it this last few days and the warnings disappeared, I promise this is not homeopathy :)

BTW, the logs I just posted are generated with this:

```yaml
sensor:
  - platform: simulated
    name: sim
    scan_interval: 1

  - platform: random
    name: rand
    minimum: 10
    maximum: 100
    scan_interval: 1

  - platform: random
    name: rand_2
    minimum: 5
    maximum: 50
    scan_interval: 1

  - platform: template
    sensors:
      results:
        value_template: '{{ (states.sensor.sim.state | float) * (states.sensor.rand.state | float) * (states.sensor.rand_2.state | float) }}'
        unit_of_measurement: "W"
        entity_id:
          - sensor.sim
          - sensor.rand
          - sensor.rand_2

  - platform: statistics
    name: stats_1
    entity_id: sensor.results
    sampling_size: 100
    max_age:
      minutes: 2
```

In more chaotic scenarios, with multiple stats sensors using template sensors depending on multiple sensors, the event bus gets stressed for sure and in that moment is when the core warning appears. 

I've not been capable of reproducing it in the dev machine as requested, but it appeared naturally at random times. My guest was to try to reduce the number of these scheduled callbacks.